### PR TITLE
Implement more validation for inline array element access and conversion scenarios

### PR DESCRIPTION
**Pull request by AlekseyTs**
Relates to test plan https://github.com/dotnet/roslyn/issues/67826

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler For the second review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler For the second review

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs [hunk 1]
```diff
@@ -768,12 +770,180 @@ void verify(ModuleSymbol m)
...
+}
+";
+            var comp = CreateCompilation(src + InlineArrayAttributeDefinition);
+            CompileAndVerify(comp, symbolValidator: verify, sourceSymbolValidator: verify).VerifyDiagnostics();
+
+            void verify(ModuleSymbol m)
+            {
+                var buffer = m.GlobalNamespace.GetTypeMember("Buffer");
+
+                Assert.True(buffer.HasInlineArrayAttribute(out int length));
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:797]: >Assert.True(buffer.HasInlineArrayAttribute(out int length)); [](http://example.com/codeflow?start=16&length=60)

Why is the type with two fields recognized as a valid inline array type?

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs [hunk 2]
```diff
@@ -2469,31 +2469,37 @@ ITypeSymbol ITypeSymbolInternal.GetITypeSymbol()
...
                         {
-                            elementType = field.TypeWithAnnotations;
+                            elementField = field;
                         }
                     }
                 }
             }
 
-            return elementType;
+            if (elementField is not null && elementField.ContainingType.IsGenericType)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs:2497]: >elementField.ContainingType.IsGenericType [](http://example.com/codeflow?start=44&length=41)

Are we testing nested types, with and without generic type parameters?

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:797]: > Why is the type with two fields recognized as a valid inline array type?

It is not. This API is not determining whether the type is a valid inline array type. It only indicates whether it has a valid attribute. The final check is performed by ```TryGetInlineArrayElementField```, which returns the only field

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs:2497]: > Are we testing nested types, with and without generic type parameters?

We are not at the moment, but ```IsGenericType``` checks covers the scenario. I will add a test.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs:2497]: Added tests.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs [hunk 3]
```diff
@@ -768,12 +770,317 @@ void verify(ModuleSymbol m)
...
+            var src = @"
+
+public class Enclosing<T>
+{
+    [System.Runtime.CompilerServices.InlineArray(10)]
+    public struct Buffer<U>
+    {
+        private T _element0;
+    }
+}
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:1040]: Consider using this type:
```
var c = new C();
c.F[2] = 1;
Console.WriteLine(c.F[2]);

class C
{
    Enclosing<int>.Buffer<string> F;
}
```

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:1072]: > Consider using this type: ...

Done

**Review by cston — APPROVED** [review]

**Pull request merged**